### ISSUE TITLE
Inconsistent logging behavior when running concurrent shell commands

### ISSUE THREAD

**Issue by yhshin11** [PrefectHQ/prefect-shell#32]
Hello all. I've been trying to implement a very standard command-line workflow with Prefect 2.0. Here's what I want to do:
- Process many files with the same shell script in parallel
- Track progress of overall processing and individual files using Orion UI
  - Overall progress: Number of current running/failed/completed tasks
  - Individual progress: View stdout/stderr logs of tasks that are complete, or are *currently running*

I'm having trouble with the part in emphasis, namely viewing the task logs *while the tasks are running*.

I've made a repo that illustrates what I'm trying to do: https://github.com/yhshin11/prefect-test/tree/979ed69c7b1a250060a1d3b004f550f8b07e0409

It contains a simple CLI script `script.py` that simply repeats an input message for N seconds.
Let's say we want to run this script with multiple input messages in parallel using `prefect-shell`.
To do this, I created a flow that calls `shell_run_command` multiple times, using the `task.submit()` API.

The output below is what what I get as output of `python run_tasks.py`.
The problem I'm trying to solve is that the log entries are only generated *after* all the tasks are complete, whereas I want to view the log entries *while the tasks are running*.

```
14:24:52.596 | INFO    | prefect.engine - Created flow run 'rough-husky' for flow 'main_flow'
14:24:52.733 | INFO    | Flow run 'rough-husky' - Created task run 'shell_run_command-7398b6ba-1' for task 'shell_run_command'
14:24:52.734 | INFO    | Flow run 'rough-husky' - Submitted task run 'shell_run_command-7398b6ba-1' for execution.
14:24:52.795 | INFO    | Flow run 'rough-husky' - Created task run 'shell_run_command-7398b6ba-0' for task 'shell_run_command'
14:24:52.796 | INFO    | Flow run 'rough-husky' - Submitted task run 'shell_run_command-7398b6ba-0' for execution.
14:24:53.172 | INFO    | Flow run 'rough-husky' - Created task run 'shell_run_command-7398b6ba-2' for task 'shell_run_command'
14:24:53.172 | INFO    | Flow run 'rough-husky' - Submitted task run 'shell_run_command-7398b6ba-2' for execution.
14:24:53.624 | INFO    | Flow run 'rough-husky' - Created task run 'shell_run_command-7398b6ba-3' for task 'shell_run_command'
14:24:53.625 | INFO    | Flow run 'rough-husky' - Submitted task run 'shell_run_command-7398b6ba-3' for execution.
14:25:03.379 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 0 seconds... Repeating message: woof
Been running task for 1 seconds... Repeating message: woof
Been running task for 2 seconds... Repeating message: woof
Been running task for 3 seconds... Repeating message: woof
Been running task for 4 seconds... Repeating message: woof
Been running task for 5 seconds... Repeating message: woof
Been running task for 6 seconds... Repeating message: woof
Been running task for 7 seconds... Repeating message: woof
Been running task for 8 seconds... Repeating message: woof
Been running task for 9 seconds... Repeating message: woof
Finished running task!

14:25:03.764 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 0 seconds... Repeating message: moo
Been running task for 1 seconds... Repeating message: moo
Been running task for 2 seconds... Repeating message: moo
Been running task for 3 seconds... Repeating message: moo
Been running task for 4 seconds... Repeating message: moo
Been running task for 5 seconds... Repeating message: moo
Been running task for 6 seconds... Repeating message: moo
Been running task for 7 seconds... Repeating message: moo
Been running task for 8 seconds... Repeating message: moo
Been running task for 9 seconds... Repeating message: moo
Finished running task!

14:25:03.945 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 0 seconds... Repeating message: quack
Been running task for 1 seconds... Repeating message: quack
Been running task for 2 seconds... Repeating message: quack
Been running task for 3 seconds... Repeating message: quack
Been running task for 4 seconds... Repeating message: quack
Been running task for 5 seconds... Repeating message: quack
Been running task for 6 seconds... Repeating message: quack
Been running task for 7 seconds... Repeating message: quack
Been running task for 8 seconds... Repeating message: quack
Been running task for 9 seconds... Repeating message: quack
Finished running task!

14:25:04.760 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Finished in state Completed()
14:25:04.786 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 0 seconds... Repeating message: meow
Been running task for 1 seconds... Repeating message: meow
Been running task for 2 seconds... Repeating message: meow
Been running task for 3 seconds... Repeating message: meow
Been running task for 4 seconds... Repeating message: meow
Been running task for 5 seconds... Repeating message: meow
Been running task for 6 seconds... Repeating message: meow
Been running task for 7 seconds... Repeating message: meow
Been running task for 8 seconds... Repeating message: meow
Been running task for 9 seconds... Repeating message: meow
Finished running task!

14:25:04.929 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Finished in state Completed()
14:25:05.205 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Finished in state Completed()
14:25:05.260 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Finished in state Completed()
14:25:05.288 | INFO    | Flow run 'rough-husky' - Finished in state Completed('All states completed.')
```

Weirdly enough, when I run the exact same script using the VS Code python debugger, I get the following output (which is the behavior I want):

```
14:51:01.863 | INFO    | prefect.engine - Created flow run 'chirpy-piculet' for flow 'main_flow'
14:51:02.103 | INFO    | Flow run 'chirpy-piculet' - Created task run 'shell_run_command-7398b6ba-0' for task 'shell_run_command'
14:51:02.105 | INFO    | Flow run 'chirpy-piculet' - Submitted task run 'shell_run_command-7398b6ba-0' for execution.
14:51:02.415 | INFO    | Flow run 'chirpy-piculet' - Created task run 'shell_run_command-7398b6ba-1' for task 'shell_run_command'
14:51:02.416 | INFO    | Flow run 'chirpy-piculet' - Submitted task run 'shell_run_command-7398b6ba-1' for execution.
14:51:03.686 | INFO    | Flow run 'chirpy-piculet' - Created task run 'shell_run_command-7398b6ba-3' for task 'shell_run_command'
14:51:03.687 | INFO    | Flow run 'chirpy-piculet' - Submitted task run 'shell_run_command-7398b6ba-3' for execution.
14:51:05.463 | INFO    | Flow run 'chirpy-piculet' - Created task run 'shell_run_command-7398b6ba-2' for task 'shell_run_command'
14:51:05.464 | INFO    | Flow run 'chirpy-piculet' - Submitted task run 'shell_run_command-7398b6ba-2' for execution.
14:51:07.697 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 0 seconds... Repeating message: moo

14:51:08.698 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 1 seconds... Repeating message: moo

14:51:08.729 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 0 seconds... Repeating message: woof

14:51:09.699 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 2 seconds... Repeating message: moo

14:51:09.730 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 1 seconds... Repeating message: woof

14:51:09.774 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 0 seconds... Repeating message: meow

14:51:10.645 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 0 seconds... Repeating message: quack

14:51:10.700 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 3 seconds... Repeating message: moo

14:51:10.731 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 2 seconds... Repeating message: woof

14:51:10.776 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 1 seconds... Repeating message: meow

14:51:11.646 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 1 seconds... Repeating message: quack

14:51:11.702 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 4 seconds... Repeating message: moo

14:51:11.733 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 3 seconds... Repeating message: woof

14:51:11.777 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 2 seconds... Repeating message: meow

14:51:12.647 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 2 seconds... Repeating message: quack

14:51:12.703 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 5 seconds... Repeating message: moo

14:51:12.733 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 4 seconds... Repeating message: woof

14:51:12.778 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 3 seconds... Repeating message: meow

14:51:13.648 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 3 seconds... Repeating message: quack

14:51:13.704 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 6 seconds... Repeating message: moo

14:51:13.734 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 5 seconds... Repeating message: woof

14:51:13.779 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 4 seconds... Repeating message: meow

14:51:14.649 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 4 seconds... Repeating message: quack

14:51:14.705 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 7 seconds... Repeating message: moo

14:51:14.735 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 6 seconds... Repeating message: woof

14:51:14.780 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 5 seconds... Repeating message: meow

14:51:15.650 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 5 seconds... Repeating message: quack

14:51:15.706 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 8 seconds... Repeating message: moo

14:51:15.736 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 7 seconds... Repeating message: woof

14:51:15.781 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 6 seconds... Repeating message: meow

14:51:16.651 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 6 seconds... Repeating message: quack

14:51:16.707 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Been running task for 9 seconds... Repeating message: moo
Finished running task!

14:51:16.737 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 8 seconds... Repeating message: woof

14:51:16.782 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 7 seconds... Repeating message: meow

14:51:17.652 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 7 seconds... Repeating message: quack

14:51:17.738 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Been running task for 9 seconds... Repeating message: woof
Finished running task!

14:51:17.783 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 8 seconds... Repeating message: meow

14:51:18.653 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 8 seconds... Repeating message: quack

14:51:18.784 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Been running task for 9 seconds... Repeating message: meow
Finished running task!

14:51:19.654 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Been running task for 9 seconds... Repeating message: quack
Finished running task!

14:51:20.542 | INFO    | Task run 'shell_run_command-7398b6ba-0' - Finished in state Completed()
14:51:22.885 | INFO    | Task run 'shell_run_command-7398b6ba-1' - Finished in state Completed()
14:51:23.675 | INFO    | Task run 'shell_run_command-7398b6ba-3' - Finished in state Completed()
14:51:24.014 | INFO    | Task run 'shell_run_command-7398b6ba-2' - Finished in state Completed()
14:51:24.063 | INFO    | Flow run 'chirpy-piculet' - Finished in state Completed('All states completed.')
```

I suspect this has something to do with how `TextReceiveStream(process.stdout)` works in `anyio` but I'm not familiar enough with the library to get the behavior I want.
Any ideas?

Thanks in advance for your help!

**Comment by ahuang11** [comment]
Hi @yhshin11, just wanted to confirm whether or not you found a solution?

**Comment by madkinsz** [comment]
You may want to adopt the streaming utilities from the core library https://github.com/PrefectHQ/prefect/blob/main/src/prefect/utilities/processutils.py

**Comment by yhshin11** [comment]
@ahuang11 Apologies. I meant to leave a comment, but got sidetracked.

TL;DR `print(text, flush=True)` ensures that stdout is flushed after each print statement, which ensures the incremental logging behavior that I wanted. [Relevant fix](https://github.com/yhshin11/prefect-test/commit/22f209b9be66dddd724ad6d980331ad5dab8a588) I'm not sure why running in VS Code debugger shows different behavior, but that's not that important to me.

I am still not getting the logging behavior I want when using `DaskTaskRunner`, but that seems to be related to [an ongoing issue](https://github.com/PrefectHQ/prefect/issues/5850) with dask logging in Prefect 2.0.

@madkinsz Thanks very much for the information. That looks very interesting. Is there a reason one might prefer `open_process` from the Prefect core library over `shell_run_command`?

Is the difference mainly what's described in `processutils.py`?

```
    - Support for Windows command joining
    - Termination of the process on exception during yield
    - Forced cleanup of process resources during cancellation
```

**Comment by madkinsz** [comment]
Oh those are pretty low level, I wouldn't recommend using them directly in your workflows. I was suggesting that the pattern for streaming contents implemented there may be useful in this library.
